text/chapter10.md

```diff
@@ -391,416 +437,935 @@ foreign import isUndefined :: forall a. Undefined a -> Boolean
+In our foreign JavaScript module, we can define `unsafeHead` as follows:
 
-The `functions` package defines similar type constructors for function arities from 0 to 10.
+```javascript
+exports.unsafeHead = arr => {
+  if (arr.length) {
+    return arr[0];
+  } else {
+    throw new Error('unsafeHead: empty array');
+  }
+};
+```
 
-We can create a function of two arguments by using the `mkFn2` function, as follows:
+## Exercises
+1. (Medium) Given a record that represents a quadratic polynomial `a*x^2 + b*x + c = 0`:
+    ```hs
+    type Quadratic = {
+      a :: Number,
+      b :: Number,
+      c :: Number
+    }
+    ```
+    Write a JavaScript function `quadraticRootsImpl` and a wrapper `quadraticRoots :: Quadratic -> Pair Complex` that uses the quadratic formula to find the roots of this polynomial. Return the two roots as a `Pair` of `Complex` numbers. *Hint:* Use the `quadraticRoots` wrapper to pass a constructor for `Pair` to `quadraticRootsImpl`.
 
-```haskell
-import Data.Function.Uncurried
+## Using Type Class Member Functions
 
-uncurriedAdd :: Fn2 Int Int Int
-uncurriedAdd = mkFn2 \n m -> m + n
-```
+Just like our earlier guide on passing the `Maybe` constructor over FFI, this is another case of writing PureScript that calls JavaScript, which in turn calls PureScript functions again. Here we will explore how to pass type class member functions over the FFI.
 
-and we can apply a function of two arguments by using the `runFn2` function:
+Here is one way to give JavaScript access PureScript's `show` so that the foreign code has a way to generate a `String` from any `Show`able type.
 
-```haskell
-> runFn2 uncurriedAdd 3 10
-13
+We start with writing a foreign JavaScript function which expects the appropriate instance of `show` to match the type of `x`.
+
+```js
+exports.boldImpl = show => x =>
+  show(x).toUpperCase() + "!!!";
 ```
 
-The key here is that the compiler _inlines_ the `mkFn2` and `runFn2` functions whenever they are fully applied. The result is that the generated code is very compact:
+Then we write the matching signature:
+```hs
+foreign import boldImpl :: forall a. (a -> String) -> a -> String
+```
 
-```javascript
-var uncurriedAdd = function (n, m) {
-    return m + n | 0;
-};
+and a wrapper function that passes the correct instance of `show`:
+```hs
+boldWrap :: forall a. Show a => a -> String
+boldWrap x = boldImpl show x
 ```
 
-For contrast, here is a traditional curried function:
+We can then call the wrapper:
+```text
+$ spago repl
 
-```haskell
-curriedAdd :: Int -> Int -> Int
-curriedAdd n m = m + n
+> import Test.Examples
+> import Data.Tuple
+> boldWrap (Tuple 1 "Hat")
+"(TUPLE 1 \"HAT\")!!!"
 ```
 
-and the resulting generated code, which is less compact due to the nested functions:
+Another option is to include the type class constraint (`Show a =>`) in the foreign signature, rather than only in the wrapping function.
```

Comment:
Is this approach more likely to break with dead-code elimination? If a type class constraint is declared, will member functions for the discovered types be preserved, or must these member functions actually be used in PS code?

Reply:
By the time we get to dead code elimination, type class constraints look like regular functions, so from that perspective it’s fine. However, this isn’t recommended because people tend to rely on the runtime representation of type class dictionaries when they do it, and that’s meant to be internal (so that we can change it in order to generate better code). We will be deprecating this in 0.14.0: https://github.com/purescript/purescript/issues/3182

Reply:
I'll remove this section then. This simplifies some things, so definitely a change for the better.